src/components/AiMarketplace/MainSection/Filter/index.js

```diff
@@ -1,50 +1,50 @@
 import React, { useState } from "react";
 import { connect } from "react-redux";
 
-import StyledExpansionPanel from "./StyledExpansionPanel.js";
+import StyledExpansionPanel from "./StyledExpansionPanel";
 import { useStylesHook } from "./styles";
-import { serviceActions } from "../../../../Redux/actionCreators/index.js";
+import { serviceActions, loaderActions } from "../../../../Redux/actionCreators";
 import {
-  defaultFilterData,
   filterParamters,
   defaultActiveFilterItem,
   defaultPaginationParameters,
-} from "../../../../utility/constants/Pagination.js";
+  generateFilterObject,
+} from "../../../../utility/constants/Pagination";
+import { LoaderContent } from "../../../../utility/constants/LoaderContent";
 
-let filterData = { ...defaultFilterData };
-
-const Filter = ({ services, pagination, updatePagination, fetchService }) => {
+const Filter = ({ services, pagination, updatePagination, fetchService, filterDataProps, startLoader, stopLoader }) => {
   const [activeFilterItem, setActiveFilterItem] = useState(defaultActiveFilterItem);
   const classes = useStylesHook();
-
-  if (services.length > 0) {
-    services.map(service => {
-      if (service.tags.length > 0) {
-        service.tags.map(tag => {
-          if (!filterData.tags.items.find(el => el.title === tag)) {
-            filterData.tags.items.push({ title: tag });
-          }
-        });
-      }
-    });
-  }
+  const filterData = {};
+  Object.entries(filterDataProps).map(([key, items]) => {
+    filterData[key] = { title: key, name: key, items };
+  });
 
   const handleActiveFilterItemChange = async event => {
+    startLoader();
     const name = event.currentTarget.name;
     const value = event.currentTarget.value;
+    const currentFilterItem = [...activeFilterItem[name]];
+    if (!currentFilterItem.includes(value)) {
+      currentFilterItem.push(value);
+    } else {
+      currentFilterItem.splice(currentFilterItem.findIndex(el => el === value), 1);
+    }
+    const currentActiveFilterData = { ...activeFilterItem, [name]: currentFilterItem };
+    const filterObj = generateFilterObject(currentActiveFilterData);
     const latestPagination = { ...pagination, ...defaultPaginationParameters, s: filterParamters[name], q: value };
     await updatePagination(latestPagination);
-    await fetchService(latestPagination);
-    setActiveFilterItem({
-      ...activeFilterItem,
-      [name]: value,
-    });
+    await fetchService(latestPagination, filterObj);
+    setActiveFilterItem(currentActiveFilterData);
+    stopLoader();
```

Comment:
All the above could be moved into an action.

Reply:
[done](https://github.com/Vivek205/refactored-dapp/pull/39/commits/711d01b309dcc071bb928d67343b72a20bb073f5)